### PR TITLE
Feature/params - implemented AudioParam::SetTargetAtTime

### PR DESCRIPTION
**Pull request by b-ma**
Hey, here is the `SetTargetAtTime` plus a couple of fixes and renaming

### CODE REVIEW

**Comment by orottier** [comment]
This looks very good! I like the new naming, and the comments you have added. I will have a more thorough look soon

File: src/param.rs [hunk 1]
```diff
@@ -29,67 +29,88 @@ pub struct AudioParamOptions {
...
     fn is_empty(&mut self) -> bool {
         self.inner.is_empty()
     }
 
-    fn retain<F>(&mut self, func: F)
-    where
-        F: Fn(&AutomationEvent) -> bool,
-    {
-        self.inner.retain(func);
+    fn unsafe_peek(&self) -> Option<&AudioParamEvent> {
```
Comment (orottier) [inline src/param.rs:92]: `unsafe` is a bit misleading here. Maybe `unsorted_peek`?

File: src/param.rs [hunk 2]
```diff
@@ -589,56 +659,155 @@ impl AudioParamProcessor {
...
+                                    self.intrisic_value = clamped;
+                                    time += dt;
+                                }
+                            }
+
+                            if !ended {
+                                // compute value for `next_block_time` so that `param.value()`
+                                // stays coherent (see. comment in `AudioParam`)
+                                // allows to properly fill k-rate within next block too
+                                let exponant =
```
Comment (orottier) [inline src/param.rs:787]: spelling: exponent

File: src/param.rs [hunk 3]
```diff
@@ -589,56 +659,155 @@ impl AudioParamProcessor {
...
+                                    -1. * ((next_block_time - start_time) / time_constant);
+                                let value = end_value + dv * exponant.exp() as f32;
+                                self.intrisic_value = value.clamp(self.min_value, self.max_value);
+                                break;
+                            } else {
+                                // setTarget has no "real" end value, compute according
+                                // to next event start time
+                                let exponant = -1. * ((end_time - start_time) / time_constant);
```
Comment (orottier) [inline src/param.rs:795]: spelling: exponent

File: src/param.rs [hunk 4]
```diff
@@ -171,7 +192,15 @@ impl AudioParam {
         self.max_value
     }
 
-    // @note - we need to test more how this behaves
+    // the choice here is to have this coherent with the first sample of
+    // the last rendered block, which means `intrisic_value must be calculated
```
Comment (orottier) [inline src/param.rs:196]: missing backtick after intrisic_value

**Review by orottier — CHANGES_REQUESTED** [review]
Some small nitpicks. After that, ready to merge!

**Comment by b-ma** [comment]
cool! thank for reviewing

**Pull request merged**